test/unit/graphql_test.lua

```diff
@@ -917,33 +917,40 @@ function g.test_rules_directivesAreDefined()
 end
 
 function g.test_types_isValueOfTheType_for_scalars()
+    local function isString(value)
+        return type(value) == 'string'
+    end
+
+    local function coerceString(value)
+        if value ~= nil then
+            value = tostring(value)
+            if not isString(value) then return end
+        end
+
+        return value
+    end
+
     t.assert_error(function()
         types.scalar({
             name = 'MyString',
-            description = "Custom string type",
-            serialize = tostring,
-            parseValue = tostring,
+            description = 'Custom string type',
+            serialize = coerceString,
+            parseValue = coerceString,
             parseLiteral = function(node)
-                if node.kind == 'string' then
-                    return node.value
-                end
+                return coerceString(node.value)
             end,
         })
     end)
 
     local CustomString = types.scalar({
         name = 'MyString',
-        description = "Custom string type",
-        serialize = tostring,
-        parseValue = tostring,
+        description = 'Custom string type',
+        serialize = coerceString,
+        parseValue = coerceString,
         parseLiteral = function(node)
-            if node.kind == 'string' then
-                return node.value
-            end
-        end,
-        isValueOfTheType = function(value)
-            return type(value) == 'string'
+            return coerceString(node.value)
         end,
+        isValueOfTheType = isString,
     })
     t.assert_equals(CustomString.__type, 'Scalar')
 end
```

Comment:
> standard == spec? Seems it doesn't. Could you link some tutorial or something to help understand the internals?

https://medium.com/@alizhdanov/lets-understand-graphql-scalars-3b2b016feb4a

https://atheros.ai/blog/how-to-design-graphql-custom-scalars

Yep, it's correct to say about implementation details. It's not described in spec. But there is some examples where it's implemented. I add two links. Seems yes, we should add it into comments/commit message.

Reply:
@olegrok I added links to description

Reply:
> Is it tested too? Could you point it out, I didn't find it.

Yes, of course.

This is test for `box.NULL` value (`box.NULL` will be returned by `parseValue` function):
https://github.com/tarantool/cartridge/blob/a39b622cd1ae9a1318e3c5d2698275241d2d2e45/test/integration/graphql_test.lua#L685-L696

This is test for `box.NULL` value (`box.NULL` will be returned by `parseLiteral` function):
https://github.com/tarantool/cartridge/blob/a39b622cd1ae9a1318e3c5d2698275241d2d2e45/test/integration/graphql_test.lua#L698-L709

This is test for `nil` value (it will be return from `parseLiteral`):
https://github.com/tarantool/cartridge/blob/a39b622cd1ae9a1318e3c5d2698275241d2d2e45/test/integration/graphql_test.lua#L144-L151

Reply:
> AFAICS the new JSON type is just a string. Seems it implies double serialization and some overhead. Is it a common way? I've seen another approach here. Could you say a few words about their pros and cons?

1. It's not a simple string, it's check availability to parse this string as JSON.
2. In your source impossible to use numbers or simple strings
3. We want to create Any type, but it's incompatible with GraphQL spec (but it's possible to implement). So, we decided that best way to solve this problem - create JSON type, that like as a string, but also have automatic JSON parsing

Reply:
> What cases does this patch fix in particular? I've seen the description of #1198, could you enrich it with some examples, please.

You can see tests in previous messages, they didn't work before patch